cli/cmd/uninject.go

```diff
@@ -0,0 +1,152 @@
+			}
+
+			in, err := read(args[0])
+			if err != nil {
+				return err
+			}
+
+			exitCode := transformInput(in, os.Stderr, os.Stdout, nil, resourceTransformerUninject{})
+			os.Exit(exitCode)
+			return nil
+		},
+	}
+
+	return cmd
+}
+
+func (rt resourceTransformerUninject) transform(bytes []byte, options *injectOptions) ([]byte, []injectReport, error) {
+	conf := &resourceConfig{}
+	output, reports, err := conf.parse(bytes, options, rt)
+	if output != nil || err != nil {
+		return output, reports, err
+	}
+
+	report := injectReport{
+		name: fmt.Sprintf("%s/%s", strings.ToLower(conf.meta.Kind), conf.om.Name),
+	}
+
+	// If we don't uninject anything into the pod template then output the
+	// original serialization of the original object. Otherwise, output the
+	// serialization of the modified object.
+	output = bytes
+	if conf.podSpec != nil {
+		uninjectPodSpec(conf.podSpec)
+		uninjectObjectMeta(conf.objectMeta, conf.k8sLabels)
+		var err error
+		output, err = yaml.Marshal(conf.obj)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		report.unsupportedResource = true
+	}
+
+	return output, []injectReport{report}, nil
+}
+
+func (resourceTransformerUninject) generateReport(uninjectReports []injectReport, output io.Writer) {
+	uninjected := []string{}
+	for _, r := range uninjectReports {
+		if !r.unsupportedResource {
+			uninjected = append(uninjected, r.name)
+		}
+	}
+	summary := fmt.Sprintf("Summary: %d of %d YAML document(s) uninjected", len(uninjected), len(uninjectReports))
+	output.Write([]byte(fmt.Sprintf("\n%s\n", summary)))
+
+	for _, i := range uninjected {
+		output.Write([]byte(fmt.Sprintf("  %s\n", i)))
+	}
+
+	// trailing newline to separate from kubectl output if piping
+	output.Write([]byte("\n"))
+}
+
+// Given a PodSpec, update the PodSpec in place with the sidecar
+// and init-container uninjected
+func uninjectPodSpec(t *v1.PodSpec) {
+	initContainers := []v1.Container{}
+	for _, container := range t.InitContainers {
+		if container.Name != k8s.InitContainerName {
+			initContainers = append(initContainers, container)
+		}
+	}
+	t.InitContainers = initContainers
+
+	containers := []v1.Container{}
+	for _, container := range t.Containers {
+		if container.Name != k8s.ProxyContainerName {
+			containers = append(containers, container)
+		}
+	}
+	t.Containers = containers
+
+	volumes := []v1.Volume{}
+	for _, volume := range t.Volumes {
+		if volume.Name != "linkerd-trust-anchors" && volume.Name != "linkerd-secrets" {
+			volumes = append(volumes, volume)
+		}
+	}
+	t.Volumes = volumes
+}
+
+func uninjectObjectMeta(t *metaV1.ObjectMeta, k8sLabels map[string]string) {
+	newAnnotations := make(map[string]string)
+	for key, val := range t.Annotations {
+		if key != k8s.CreatedByAnnotation && key != k8s.ProxyVersionAnnotation {
+			newAnnotations[key] = val
+		}
+	}
+	t.Annotations = newAnnotations
+
+	labels := make(map[string]string)
+	ldLabels := []string{k8s.ControllerNSLabel}
```

Comment:
In the following lines it expands that slice with the labels that are specific to the current resource kind being processed. Or were you implying that we should simply iterate through all the labels regardless of the resource kind?

Reply:
Ah, ok, that makes sense. I was testing this by running inject on a deployment, and then running uninject on a single pod from the deployment. For instance:

```
bin/linkerd install | kubectl apply -f -
bin/linkerd inject test/testdata/smoke_test.yaml | kubectl apply -f -
kubectl get po/smoke-test-gateway-85db44995d-8krxg -oyaml | bin/linkerd uninject -
```

And in the uninjected YAML, I still see:

```
  labels:
    app: smoke-test-gateway
    linkerd.io/proxy-deployment: smoke-test-gateway
    pod-template-hash: "4186005518"
```

Which makes sense, since the inject resource type was "deployment", but the uninject resource type was "pod". I can't think of a better fix for this than iterating through all of the labels, but I agree that it feels heavy handed.

Reply:
I hadn't realized how those labels propagate to pods, so the wholesale approach makes sense. I'll apply that :+1: